tc/core/polyhedral/cuda/codegen.cc

```diff
@@ -396,10 +396,46 @@ struct LdgWrapper {
   std::ostream& out_;
 };
 
+template <typename AFF>
+isl::ast_expr buildAccess(AFF access, const CodegenStatementContext& context) {
+  return context.build().access_from(access);
+}
+
+void emitAccess(isl::ast_expr access, const CodegenStatementContext& context) {
+  context.ss << access.to_C_str();
+}
+
 template <typename AFF>
 void emitAccess(AFF access, const CodegenStatementContext& context) {
+  emitAccess(buildAccess(access, context), context);
+}
+
+// Check that the given expression is an access with constant index expressions
+void checkConstantAccess(isl::ast_expr expr) {
+  auto op = expr.as<isl::ast_expr_op>();
+  auto access = op.as<isl::ast_op_access>();
+  TC_CHECK(access);
+  for (int i = 1; i < access.get_n_arg(); ++i) {
+    auto arg = access.get_arg(i);
+    TC_CHECK(arg.as<isl::ast_expr_int>());
+  }
+}
+
+// Print an access to a(n array of) register(s), checking that
+// the index expressions are constant.
+template <typename AFF>
+void emitRegisterAccess(AFF access, const CodegenStatementContext& context) {
+  auto expr = buildAccess(access, context);
+  checkConstantAccess(expr);
+  emitAccess(expr, context);
+}
+
+// Print an access to global memory, wrapping the access in an "__ldg()"
+// call if the accessed tensor is known to be read-only.
+template <typename AFF>
+void emitGlobalAccess(AFF access, const CodegenStatementContext& context) {
```

Comment:
Same question for the template: there are two call sites using the same argument type.